### ISSUE TITLE
The system of saving

### ISSUE THREAD

**Issue by ilka2** [CG-Tespy/Fungus-Slot-based-Save-System#20]
Hi. I came across some misunderstanding about how saves work in this extension.

First, I would like to know if the background corresponding to the moment of saving is saved somewhere. I want to put in the save slots, in addition to the save date, the background that was in the game at the moment, so that it would be easier to find the right save.

Secondly, I did not understand how to add my data to the saves and then count them. As I understand it, SaveDataMy (the so-called SaveData subclass) stores the data that we need to save. Data Saved My creates a list with this data. SaveLoaderMy takes a Json file and reads the data from there. But how to implement it is not clear to me. And it is not clear how my classes are implemented into the existing preservation system.

**Comment by CG-Tespy** [comment]
I'm... not entirely sure what you're getting at here. For example, it's not clear what you mean by "background". Also, this system doesn't have any class called "SaveDataMy". Please explain what you're trying to do as far as far as making custom save data goes.

**Comment by ilka2** [comment]
1. Background. When I have saved the Fungus data, I want to upload it. The download works fine, except instead of the background (what the "view" indicates) I get a gray picture.
2. All classes ending in "...My" are subclasses of the classes of the same name from the SlotBasedSystem (SaveData, DataSaver, SaveLoader). In my understanding, SaveDataMy stores user data that we need to save. DataSaverMy - creates a list with this data (which we will have to write to JSON later). SaveLoaderMy - takes a Json file and reads data from there. But it sounds beautiful in words. And how it actually interacts, I don't understand.

**Comment by CG-Tespy** [comment]
Sounds to me like I should explain how classes like SaveData and DataSaver interact. Let’s start with...

# The Process of Creating Save Data

This is centered around the GameSaver component: the head honcho of this process. It creates GameSaveData instances that hold the more specific save data types (such as FlowchartData). 

Right after creating such an instance, it passes it around to its minions (which I also call subsavers), one by one. The minions here are DataSaver subclasses. One example of that is FlowchartSaver. Their job is to gather up what they want saved and then give it to the head honcho. The head honcho then takes what it was given, adding it to the GameSaveData instance. When the subsavers' work is done, the GameSaver adds in stuff like the scene name and the last-hit Progress Marker. And with that, the GameSaveData instance is ready to be loaded, written to disk, and such. 


There’s a lot more involved in this system (such as the process of **reading** save data). But before I move onto that, please let me know if you have any questions.

**Comment by ilka2** [comment]
Yes, everything seems to be clear here. We have a GameSaver that creates a list of additional subordinates (derived from DataSaver). They're collecting some kind of data. Then we create an object from GameSaveData, an object is created where we record everything we have collected.

**Comment by CG-Tespy** [comment]
Glad to hear! Now, as for the process of loading save data... The structure is similar to the process for saving it. You have one main module tasked to do a job, and then it has its submodules do different parts of said job. In this case, the main module is GameLoader. The submodules are derived from SaveLoader.

Any questions? Sorry for taking so long to respond, by the way...

**Comment by ilka2** [comment]
I also study and work. There is very little time. I need to comprehend and try out everything that is written here in order to clarify the questions. Don't close the subject yet, please.

**Comment by ilka2** [comment]
I read it all again and looked at it. Without going into the code (because I'm a bad programmer), everything is clear to me. Next, should we create derived classes from Gamesaver and GameLoader?

**Comment by CG-Tespy** [comment]
No, you're supposed to subclass DataSaver and SaveLoader. You also need to make sure that you have those subclasses added to the GameSaver and GameLoader GameObjects, respectively. Those GameObjects are in the SaversAndLoaders prefab.

**Comment by ilka2** [comment]
Let's say. When will I need to write my own functions to save the parameters? Or is it possible to abandon your class from the "GameSaveData" class and not from "SaveData"?

**Comment by ilka2** [comment]
I created these subclasses. But I don't really understand what to do with them next. And should I inherit my classes from "DataSaver" and "SaveLoader"?

UPD: I'm sorry I've been away so long.

**Comment by CG-Tespy** [comment]
>Should I inherit my classes from DataSaver and DataLoader

Yes, that's what subclassing means.  

>What to do with them next

As I mentioned, you add them to the GameSaver and GameLoader GameObjects. To be more precise (since we're working with Unity), this means adding them as Components to those GameObjects.

Also, I feel I should mention... Reasonably-robust save systems are an intermediate-level topic as far as programming is concerned. If you don't have a coder who is at least on that level, then I suggest doing at least one of the following:

- Recruiting one such programmer
- Taking courses so you can get to that level
  - I can recommend you a couple, if you'd like

Without a coder on that level on your team, you're probably not going to be able to do what you want with this save system. Not if you need more than the built-in functionality, anyway

**Comment by ilka2** [comment]
I am the programmer in the team. And I have some knowledge. And not only in C#, but it's hard for me to work with someone else's code, because my tasks are usually simpler (the same inheritance).
I also understand that scripts are components of objects. I meant what should I do next with the subclasses. As I understand it, all my variables will be added in the form of "SaveDataItem" or "GameSaveData". That's the question, how do I make "DataSaver" pack into .json my variables and then "SaveLoader" from the file (or rather from the registry, because we use JsonUtility) wrote the data to the variable

**Comment by CG-Tespy** [comment]
DataSaver doesn't handle packing things into json. That's what the SaveWriter class handles. As for reading **from** json, that's the SaveReader class's job

**Comment by ilka2** [comment]
Sorry. I have to keep asking questions, because the problem has not been solved.

I will specify the task. I need to save a string that the user can enter in the Input Field. To simplify, I decided to save the string in an instance of SaveDataItem. If I understand correctly, I can create another class, inherit it from SaveData and store my variables there.

We have MySaver:DataSaver and MySaveLoader:SaveLoader. There you need to define CreateItems() and Load(SaveDataItem item), respectively. But I don't understand what should be in these methods? I hoped the implementation of FlowchartSaver and FlowchatLoader would help me, but not no.

Can you please explain what my classes specifically do?

**Comment by CG-Tespy** [comment]
Sorry for taking so long to respond ^^; And yes, you have the right idea with storing the string into a SaveDataItem. You are also correct about how you can inherit from SaveData and store your variables there. As for what you do with CreateItems... well, you set up a list of SaveDataItem instances and return it. In your case, since you want to save a string from an input field, you put that string into one of those instances. As for Load(InsertSaveDataSubclassHere)... it's as it says on the tin. It loads what's in the save data passed to that function.

As for what the classes specifically do... Well, their names give it away. DataSaver saves the save data, SaveLoader loads it.

**Comment by ilka2** [comment]
It seems that we have succeeded). thank you.

**Comment by ilka2** [comment]
Good afternoon. I feel really embarrassed to keep bothering you with questions, but I genuinely can't figure this out on my own. And yes, you're right—my programming knowledge is quite limited. I know some things in theory, but since I don't practice much, applying that knowledge in real scenarios is really challenging for me.

I want to create my own class that would replace SaveDataItem. I'll call this class MySaveDataItem. I have a derived class MySaver : DataSaver, which contains a method IList<SaveDataItem> CreateItems(). After looking at its implementation, I realized that this method specifically expects SaveDataItem and nothing else.

Therefore, I can't inherit MySaveDataItem from SaveData, as suggested in "Getting Started," because the classes aren't connected in any way.

So, I decided to inherit MySaveDataItem from SaveDataItem. However, my attempt to create a List<SaveDataItem> items with an element new MySaveDataItem() also failed because CreateItems() still only accepts SaveDataItem, and it doesn't recognize anything added in the derived class.

I didn't even consider type casting. Please save me once again, and I hope this will be the last time. Thank you!

**Comment by CG-Tespy** [comment]
In regards to having a hard time applying the knowledge in real scenarios being real hard...
Considering how long you've been a programmer, it's best that you take some courses. Feels like you're improving at a much slower pace than you should be. I can recommend some courses, if you'd like.

In regards to being unable to return a list of MySaveDataItem from CreateItems()...
Given how polymorphism works, add the MySaveDataItem instances to a list of SaveDataItem. If the former inherits from the latter, then it should work

**Comment by ilka2** [comment]
Sorry for the long answer. Yes, I will be grateful if you share the courses. I can't promise that I will be able to pass them (because of the language barrier), but I will try.

I couldn't do what I wanted to do. Right now my code looks like this:

A script that is a template for saving data:
```
public class MySaveDataItem : SaveDataItem
{
    public Vector3 camPos;
}
```

The script for saving the game:
```
public class MySaver : DataSaver
{
    public override IList<SaveDataItem> CreateItems()
    {
        Vector3 camPosSave = GameObjectExtension.Find("Main Camera").transform.position;

        List<SaveDataItem> items = new List<SaveDataItem>();
        items.Add(new MySaveDataItem() { camPos = camPosSave });

        return items;
    }
}
```

The download script:
```
public class MySaveLoader : SaveLoader
{
    public override bool Load(SaveDataItem item)
    {
        if (item == null)
        {
            Debug.LogError("Item is null.");
            return false;
        }

        if (item.Type == typeof(MySaveDataItem).FullName)
        {
            MySaveDataItem myItem = item as MySaveDataItem;
            Camera.main.transform.position = myItem.camPos;
            Debug.Log("Data loaded successfully: " + myItem.camPos);
            return true;
        }
        else
        {
            return false;
        }
    }
}
```
I also added the Type property to the "SaveDataItem". And in the SaveDataItem constructor, I refer to the property:
`public SaveDataItem() { Type = GetType().FullName; }`

Everything is working. But before restarting the game. What could be the reason for this?

P.S.  The upload script does not find an item with the type "MySaveDataItem".

**Comment by CG-Tespy** [comment]
Here are some courses from [GameDevTV](https://gamedev.tv/) that can help you a lot:
- Complete Unity 2D Developer: Create Your Own 2D Games Using Unity C#
- Complete Unity 3D Developer: Design & Develop Games in Unity 6 using C#

[Zenva](https://academy.zenva.com/) is another really good learning platform. If you want to give them a try, I recommend their Unity Associate Programmer Certification Pathway. 

Every few months, [HumbleBundle](https://humblebundle.com/) has bundles for GDTV or Zenva courses. Might be good to keep your eyes peeled for those.

As for the new issues... what did you mean by "Everything is working but before restarting the game"? As well as the DataSaver not finding MySaveDataItem instances.

**Comment by ilka2** [comment]
I'm running a project in the Unity editor. Saving the game. Then I load the game through the menu provided in the asset. Everything is working. Then I stop the game. I'm starting it again (also in the editor). I'm trying to load the same save, but MySaveLoader returns false because there is not a single item whose type would be == "MySaveDataItem". I can't imagine how this can happen yet, because before I stop the "show" of the game, everything loads perfectly.

**Comment by CG-Tespy** [comment]
It sounds like the MySaveDataItems weren't properly written to the save file. Can you post the contents of the file on pastebin or something and show it here? It should be in pretty-printed json by default, so...

**Comment by ilka2** [comment]
Hi. Finally got to the game. I wish I'd thought of looking at the file myself. For some reason, I thought that everything was stored through PlayerPrefs, although I myself had viewed these files once...

In general, I share everything I have. And I'll watch it myself if I have time, maybe I'll figure it out myself and I won't need to waste your time. I've already spent more than I expected. I'm sorry.

I've never used pastebin. I hope I did everything right. https://pastebin.com/aiH58e2z

P.S.
I can see that my data is not here. Hence the two questions
1. During the project's debug, is the data stored in one of the game objects?
2. If not, how can I save my data type through the SaveDataItem wrapper so that it is saved correctly. How I did it can be read and seen above (I attached the code).

**Comment by CG-Tespy** [comment]
Unfortunately, I can't view that paste; it's marked as private. With that said, did you make sure that your MySaveDataItem class is set to be serializable? If not, that would explain why it wasn't written to the file

**Comment by ilka2** [comment]
Make it public, even though there's really nothing to see there.

I don't think I set the serialization. I'll try in the morning, thanks.

P.S. I returned to the c++ programming courses. I had to take a break due to lack of time. I hope I'll improve my knowledge there now and write better games for unity :/
Although some topics, like library creation and everything related to links, are killing me. Fortunately, c# is much kinder to users

**Comment by ilka2** [comment]
I think I've been able to fix everything. In case I'm not the only sufferer, I'll try to describe what I did. The code that I provided before is no longer relevant. I have removed the changes in "SaveDataItem". All the fixes will be in my scripts only.

MySaveSataItem:
```
using CGTUnity.Fungus.SaveSystem;
using UnityEngine;

[System.Serializable]
public class MySaveDataItem : SaveDataItem
{
    [SerializeField] protected Vector3 camPos;
    public Vector3 CamPos { get { return camPos; } set { camPos = value; } }
}
```

MySaver:
```
using CGTUnity.Fungus.SaveSystem;
using UnityEngine;

[System.Serializable]
public class MySaveDataItem : SaveDataItem
{
    [SerializeField] protected Vector3 camPos;
    public Vector3 CamPos { get { return camPos; } set { camPos = value; } }
}
```

MySaveLoader:
```
using CGTUnity.Fungus.SaveSystem;
using System.Collections;
using System.Collections.Generic;
using TMPro;
using UnityEngine;

public class MySaveLoader : SaveLoader
{
    public override bool Load(SaveDataItem item)
    {
        if (item == null)
        {
            Debug.LogError("Item is null.");
            return false;
        }

        if (item.DataType == typeof(MySaveDataItem).Name)
        {
            MySaveDataItem myItem = JsonUtility.FromJson<MySaveDataItem>(item.Data);

            if (myItem == null)
            {
                Debug.LogError("Failed to deserialize MySaveDataItem.");
                return false;
            }

            Camera.main.transform.position = myItem.CamPos;
            Debug.Log("Data loaded successfully: " + myItem.CamPos);
            return true;
        }
        else
        {
            return false;
        }
    }
}
```

The problem was that I didn't call To Save Data Item() in MySaver, which converts the data to JSON.

P.S. Thanks to CG-Tespy for your patience. Thanks to DeepSeek for the help. Peace to your home.
